File: 11-graph/00200-Number_Of_Islands/00200-number_of_islands.java
// 200: Number Of Islands
// https://leetcode.com/problems/number-of-islands/


class Solution {
    private void dfs(char[][] grid, int i, int j) {
        if (
            i < 0 ||
            j < 0 ||
            i >= grid.length ||
            j >= grid[0].length ||
            grid[i][j] == '0'
        ) {
            return;
        }

        grid[i][j] = '0';

        dfs(grid, i + 1, j);
        dfs(grid, i, j + 1);
        dfs(grid, i - 1, j);
        dfs(grid, i, j - 1);
    }


    // SOLUTION
    public int numIslands(char[][] grid) {
        int result = 0;

        for (int i = 0; i < grid.length; i++) {
            for (int j = 0; j < grid[0].length; j++) {
                if (grid[i][j] == '1') {
                    dfs(grid, i, j);
                    result++;
                }
            }
        }

        return result;
    }

    public static void main(String[] args) {
        Solution o = new Solution();

        // INPUT
        char[][] grid = {
            {'1','1','1','1','0'},
            {'1','1','0','1','0'},
            {'1','1','0','0','0'},
            {'0','0','0','0','0'}
        };

        // OUTPUT
        var result = o.numIslands(grid);
        System.out.println(result);
    }
}

File: 11-graph/00200-Number_Of_Islands/00200-number_of_islands.py
# 200: Number Of Islands
# https://leetcode.com/problems/number-of-islands/

from typing import List


class Solution:
    # SOLUTION
    def numIslands(self, grid: List[List[str]]) -> int:
        if not grid or not grid[0]:
            return 0

        result = 0
        visit = set()
        rows, cols = len(grid), len(grid[0])

        def dfs(i, j):
            if (
                i not in range(rows)
                or j not in range(cols)
                or grid[i][j] == "0"
                or (i, j) in visit
            ):
                return

            visit.add((i, j))
            dfs(i + 1, j);
            dfs(i, j + 1);
            dfs(i - 1, j);
            dfs(i, j - 1);

        for i in range(rows):
            for j in range(cols):
                if grid[i][j] == "1" and (i, j) not in visit:
                    result += 1
                    dfs(i, j)

        return result


if __name__ == "__main__":
    o = Solution()

    # INPUT
    grid =[
        ['1','1','1','1','0'],
        ['1','1','0','1','0'],
        ['1','1','0','0','0'],
        ['0','0','0','0','0']
    ]

    # OUTPUT
    result = o.numIslands(grid)
    print(result)
